### PR TITLE
Set defence tokens to the highest state that is locked.

### PR DESCRIPTION
**Pull request by tomprince**
If no states are locked, the token is removed

### CODE REVIEW

**Comment by iakona** [comment]
what is the motivation behind this change?

**Comment by tomprince** [comment]
For Vital Strength, for example, you can set a defense token locked in state 3, and time-passes will automatically reset it to that state.

File: global-script.lua [hunk 1]
```diff
@@ -1989,7 +1989,23 @@ function handlePiece(object, depth)
     elseif object.getName() == "Blight" then
         object = resetPiece(object, Vector(0,180,0), depth)
     elseif string.sub(object.getName(),-7) == "Defence" then
-        if object.getLock() == false then object.destruct() end
+        -- Reset defense tokens to highest locked state.
+        -- Lock status is linked tied to a specific state.
+        local state = 21
```
Comment (iakona) [inline global-script.lua:1994]: instead of a hardcoded magic number, you should use the length of `getStates()` + 1

File: global-script.lua [hunk 2]
```diff
@@ -1989,7 +1989,23 @@ function handlePiece(object, depth)
...
+        local searching = true
+
+        if object.getStateId() ~= state then
+            object = object.setState(state)
+        end
+        while searching  and state >= 1 do
```
Comment (iakona) [inline global-script.lua:2000]: you have an extra space on this line

File: global-script.lua [hunk 3]
```diff
@@ -2120,10 +2120,24 @@ function handlePiece(object, offset)
...
+        while searching and state >= 1 do
+            if object.getLock() == true then
+                searching = false
+            else
+                state = state - 1
+                if state >= 1 then object = object.setState(state) end
+            end
         end
+        if searching then object.destruct() end
+        object = nil
```
Comment (iakona) [inline global-script.lua:2140]: object should only be set to nil in the case that it's destructed

**Comment by iakona** [comment]
so generally speaking shouldn't this be set to the lowest locked state? players are adding defense ontop vital so you're going to be looking for a low number

**Comment by tomprince** [comment]
I don't think it matters in practice, as only one state would be locked. I don't have a good intuition for what should be done if there are multiple locked states. I'm inclined to leave it as is.

**Pull request closed without merging**